drivers/gpio/gpio_stmpe1600.c

```diff
@@ -0,0 +1,297 @@
+#define REG_SYS_CTRL    0x03    /* Reset and interrupt control */
+#define REG_IEGPIOR_LSB 0x08    /* GPIO interrupt enable register */
+#define REG_IEGPIOR_MSB 0x09
+#define REG_ISGPIOR_LSB 0x0A    /* GPIO interrupt status register */
+#define REG_ISGPIOR_MSB 0x0B
+#define REG_GPMR_LSB    0x10    /* GPIO monitor pin state register */
+#define REG_GPMR_MSB    0x11
+#define REG_GPSR_LSB    0x12    /* GPIO set pin state register */
+#define REG_GPSR_MSB    0x13
+#define REG_GPDR_LSB    0x14    /* GPIO set pin direction register */
+#define REG_GPDR_MSB    0x15
+#define REG_GPPIR_LSB   0x16    /* GPIO polarity inversion register */
+#define REG_GPPIR_MSB   0x17
+
+
+/** Configuration data */
+struct stmpe1600_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+
+	/** The master I2C device's name */
+	const char *const i2c_master_dev_name;
+
+	/** The slave address of the chip */
+	uint16_t i2c_slave_addr;
+};
+
+/** Runtime driver data */
+struct stmpe1600_drvdata {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+
+	/** Master I2C device */
+	const struct device *i2c_master;
+
+	/** Driver lock */
+	struct k_sem lock;
+};
+
+static int write_reg16(const struct device *dev, uint8_t reg, uint16_t value)
+{
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct device *i2c_master = drvdata->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	int ret = 0;
+	uint16_t transfer_data = sys_cpu_to_le16(value);
+
+	LOG_DBG("STMPE1600[0x%02X]: write REG[0x%02X..0x%02X] = %04x",
+		i2c_addr, reg, reg + 1, value);
+
+	ret = i2c_burst_write(
+		i2c_master, i2c_addr, reg,
+		(uint8_t *)&transfer_data, sizeof(transfer_data)
+		);
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: write error REG[0x%02X..0x%02X]: %d",
+			i2c_addr, reg, reg + 1, ret);
+	}
+	return ret;
+}
+
+static int read_reg16(const struct device *dev, uint8_t reg, uint16_t *value)
+{
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct device *i2c_master = drvdata->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	int ret = 0;
+	uint16_t transfer_data = 0;
+
+	LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X]", i2c_addr, reg, reg + 1);
+
+	ret = i2c_burst_read(
+		i2c_master, i2c_addr, reg,
+		(uint8_t *) &transfer_data, sizeof(transfer_data)
+		);
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: read error REG[0x%02X..0x%02X]: %d",
+			i2c_addr, reg, reg + 1, ret);
+	} else {
+		*value = sys_le16_to_cpu(transfer_data);
+		LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X] => %04x",
+			i2c_addr, reg, reg + 1, *value);
+	}
+	return ret;
+}
+
+static int set_pin_dir(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	uint16_t reg_dir = 0, reg_state = 0;
```

Comment:
The values of GPSR, GPDR registers need to be cached by the driver and `reg_dir`, `reg_state` should be initialized with those cached values. Otherwise every time the function is called you're going to configure correctly `pin` but set all the remaining pins as an input. I.e. I2C transaction will reprogram full content of GPDR register, not a single bit.